File: typescript/examples/calculator/handlers/calculator-handlers.ts
import { defaultRegistry } from "@aps-kit/typescript";
import { strict as assert } from "node:assert";

class Calculator {
  value = 0;
  add(a: number, b: number) {
    this.value = a + b;
  }
  subtract(a: number, b: number) {
    this.value = a - b;
  }
}

defaultRegistry.step("a fresh calculator", (world) => {
  world.calc = new Calculator();
});

defaultRegistry.step("I add <a> and <b>", (world, ex) => {
  (world.calc as Calculator).add(parseInt(ex.a, 10), parseInt(ex.b, 10));
});

defaultRegistry.step("I subtract <b> from <a>", (world, ex) => {
  (world.calc as Calculator).subtract(parseInt(ex.a, 10), parseInt(ex.b, 10));
});

defaultRegistry.step("the result is <sum>", (world, ex) => {
  assert.equal((world.calc as Calculator).value, parseInt(ex.sum, 10));
});

defaultRegistry.step("the result is <diff>", (world, ex) => {
  assert.equal((world.calc as Calculator).value, parseInt(ex.diff, 10));
});

File: python/examples/calculator/handlers/calculator_handlers.py
from aps_kit.registry import default_registry as registry


class Calculator:
    def __init__(self) -> None:
        self.value = 0

    def add(self, a: int, b: int) -> None:
        self.value = a + b

    def subtract(self, a: int, b: int) -> None:
        self.value = a - b


@registry.step("a fresh calculator")
def _(world, _ex):
    world["calc"] = Calculator()


@registry.step("I add <a> and <b>")
def _(world, ex):
    world["calc"].add(int(ex["a"]), int(ex["b"]))


@registry.step("I subtract <b> from <a>")
def _(world, ex):
    world["calc"].subtract(int(ex["a"]), int(ex["b"]))


@registry.step("the result is <sum>")
def _(world, ex):
    assert world["calc"].value == int(ex["sum"]), (
        f"expected {ex['sum']}, got {world['calc'].value}"
    )


@registry.step("the result is <diff>")
def _(world, ex):
    assert world["calc"].value == int(ex["diff"]), (
        f"expected {ex['diff']}, got {world['calc'].value}"
    )
